binderhub/repoproviders.py

```diff
@@ -356,3 +364,54 @@ def get_resolved_ref(self):
 
     def get_build_slug(self):
         return '{user}-{repo}'.format(user=self.user, repo=self.repo)
+
+
+class GitHubGistRepoProvider(GitHubRepoProvider):
+    """GitHub gist provider.
+
+    Users must provide a spec that matches the following form (similar to github)
+
+    <username>/<gist-id>[/<ref>]
+    """
+
+    def __init__(self, *args, **kwargs):
+        # We dont need to initialize entirely the same as github
+        super(RepoProvider, self).__init__(*args, **kwargs)
+        parts = self.spec.split('/')
+        self.user, self.gist_id, *_ = parts
+        if len(parts) > 2:
+            self.unresolved_ref = parts[1]
+        else:
+            self.unresolved_ref = ''
+
+    def get_repo_url(self):
+        return f'https://gist.github.com/{self.gist_id}.git'
+
+    @gen.coroutine
+    def get_resolved_ref(self):
+        if hasattr(self, 'resolved_ref'):
+            return self.resolved_ref
+
+        api_url = f"https://api.github.com/gists/{self.gist_id}"
+        self.log.debug("Fetching %s", api_url)
+
+        resp = yield self.github_api_request(api_url)
+        if resp is None:
+            return None
+
+        ref_info = json.loads(resp.body.decode('utf-8'))
+        all_versions = [e['version'] for e in ref_info['history']]
+        if not self.unresolved_ref:
+            self.resolved_ref = all_versions[0]
+        else:
+            if self.unresolved_ref not in all_versions:
+                return None
+            else:
+                self.resolved_ref = self.unresolved_ref
+
+        return self.resolved_ref
+
+    def get_build_slug(self):
+        return self.gist_id
```

Comment:
This will coalesce multiple revisions of gists building at the same time, which isn't what we want. It should involve both the id and the resolved ref IMO.

Reply:
we dont use the resolved ref for any of the other build slugs, that gets added later

Reply:
Ah, whoops. Sorry about that.